File: src/main/java/com/flab/license/common/response/ApiResponse.java
package com.flab.license.common.response;

import org.springframework.lang.NonNull;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.flab.license.common.exception.ErrorCode;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record ApiResponse<T>(
	boolean success,
	String error,
	String message,
	T data
) {
	public static <T> ApiResponse<T> success(T data) {
		return new ApiResponse<>(true, null, null, data);
	}

	public static ApiResponse<Void> ok() {
		return new ApiResponse<>(true, null, null, null);
	}

	public static ApiResponse<Void> error(@NonNull ErrorCode errorCode) {
		return new ApiResponse<>(false, errorCode.name(), errorCode.getMessage(), null);
	}
}

File: src/main/kotlin/com/flab/license/common/response/ApiResponse.kt
package com.flab.license.common.response

import com.fasterxml.jackson.annotation.JsonInclude
import com.flab.license.common.exception.ErrorCode

@JsonInclude(JsonInclude.Include.NON_NULL)
data class ApiResponse<T>(
    val success: Boolean,
    val error: String? = null,
    val message: String? = null,
    val data: T? = null
) {
    companion object {
        fun <T> success(data: T): ApiResponse<T> = ApiResponse(success = true, data = data)

        fun ok(): ApiResponse<Unit> = ApiResponse(success = true)

        fun error(errorCode: ErrorCode): ApiResponse<Unit> = ApiResponse(
            success = false,
            error = errorCode.name,
            message = errorCode.message
        )
    }
}
